### ISSUE TITLE
CORS fail with query v2 compatibility API

### ISSUE THREAD

**Issue by LaXiS96** [influxdata/influxdb#19715]
__Description:__
I am noticing a strange behaviour when querying InfluxDB 1.8.3 with Flux, from an Angular web app that uses `@influxdata/influxdb-client`. In my particular case, my web app is accessible at `http://localhost:4200` while InfluxDB is hosted on another machine at `http://192.168.225.5:8086`.
The query request fails because of CORS policy, since this is an actual cross-origin request. But should it?
According to issue influxdata/influxdb#18391 and the consequent pull request, CORS should be correctly handled by InfluxDB and I supposedly should not be receiving this error:
![image](https://user-images.githubusercontent.com/8070973/95433990-fbb66380-0950-11eb-9a10-e54cd5dc4a53.png)
In the network tab, this is the situation:
![image](https://user-images.githubusercontent.com/8070973/95434233-4b952a80-0951-11eb-8202-c9a80ee62a3f.png)
HTTP pre-flight OPTIONS requests look like this:
![image](https://user-images.githubusercontent.com/8070973/95434344-6b2c5300-0951-11eb-8c87-8aa817e4864a.png)

I also tried reverse-proxying InfluxDB through nginx to bypass CORS using [this configuration](https://enable-cors.org/server_nginx.html): at first it seemed to work (the OPTIONS pre-flight request was successful) but any subsequent POST included duplicate Access-Control-* headers (both those returned by InfluxDB and those added by nginx).
Network tab:
![image](https://user-images.githubusercontent.com/8070973/95438428-8ea5cc80-0956-11eb-8c78-9f4131820548.png)
Bypassed pre-flight OPTIONS request:
![image](https://user-images.githubusercontent.com/8070973/95438498-a8471400-0956-11eb-94b8-c37bcc51cd0f.png)
POST response, including duplicate Access-Control-* headers:
![image](https://user-images.githubusercontent.com/8070973/95438695-efcda000-0956-11eb-975d-843a0684acf2.png)

This tells me that CORS is indeed handled correctly by InfluxDB, if not for pre-flight requests. That is, the OPTIONS requests on `/api/v2/query` should not return 405 Method Not Allowed but a 2xx success response including the relevant Access-Control-* headers.

__Steps to reproduce:__
You could try creating an Angular web app using `@influxdata/influxdb-client`, but a simple web page including a script that launches an XMLHttpRequest to InfluxDB `/api/v2/query` should be sufficient to trigger the browser's CORS checks.

__Expected behavior:__
Calls to `/api/v2/query` should respond correctly to OPTIONS requests and not fail CORS policy checks.

__Actual behavior:__
OPTIONS pre-flight request to `/api/v2/query` responds with 405 Method Not Allowed and fails CORS policy checks.

__Environment info:__

* System info: Linux 4.19.15-1.ph3-esx x86_64
* InfluxDB version: 1.8.3, Docker image influxdb:1.8 sha256:35de8cc24dfca8ad80fc93bba1bc7125240b3e27b457a1ef529aac5fb6217eef
* Other relevant environment details: Docker 18.06.1 on VMware PhotonOS 3.0

__Config:__
Only customization is `INFLUXDB_HTTP_FLUX_ENABLED=true` environment variable.

**Comment by sranka** [comment]
Thank you for reporting this issue, I can reproduce it locally. Your type of workaround is the right way to go now. Please try to adjust header specification in nginx configuration:

```
proxy_hide_header 'Access-Control-Allow-Methods';
proxy_hide_header 'Access-Control-Allow-Headers';
proxy_hide_header 'Access-Control-Allow-Origin';
add_header 'Access-Control-Allow-Origin' '*';
add_header 'Access-Control-Allow-Methods' 'POST, GET, OPTIONS, PUT, DELETE, PATCH';
add_header 'Access-Control-Allow-Headers' 'Accept, Content-Type, Content-Length, Accept-Encoding, Authorization, User-Agent';
```

The `proxy_hide_header` part gives full control of CORS to NGINX, the right CORS header values are taken from 
https://github.com/influxdata/influxdb/blob/bc4bae373852d38557ff26a35d084a9bde6d1eb4/kit/transport/http/middleware.go#L23-L33

https://github.com/influxdata/influxdb/pull/18410 contained a fix for this in compliance with the current README.MD :
```
If you are looking for the 1.x line of releases, there are branches for each minor version as well as a master-1.x branch that will contain the code for the next 1.x release. The master-1.x working branch is here. 
```
The 1.8.3 release originated in the 1.8 branch that does not contain the fix. @russorat can you please help with some clarification of where to deliver fixes for influxdb v1?

**Comment by dgnorton** [comment]
@sranka for v1, changes go into the `master-1.x` branch first and then get cherry-picked back to the desired release branch(es). Can you cherry-pick that change to the `1.8` branch and make a PR? We can get it merged for the 1.8.4 release. Thanks for the help with this issue.

**Comment by LaXiS96** [comment]
@sranka thank you for your work.
I am now using this nginx configuration as a workaround:
```
if ($request_method = 'OPTIONS') {
    add_header 'Access-Control-Allow-Origin' '*';
    add_header 'Access-Control-Allow-Methods' 'DELETE, GET, OPTIONS, POST, PUT';
    add_header 'Access-Control-Allow-Headers' 'Accept, Accept-Encoding, Authorization, Content-Length, Content-Type, X-CSRF-Token, X-HTTP-Method-Override';
    add_header 'Content-Type' 'text/plain; charset=utf-8';
    add_header 'Content-Length' 0;
    return 204;
}
```
(header values are not the same as in your comment, I took them from [here](https://github.com/influxdata/influxdb/blob/563e6c3d1a7a2790763c6289501095dbec19244e/services/httpd/handler.go#L1850))

Always rewriting all Access-Control-* headers does not work since the main problem is a 405 response to OPTIONS requests.
By bypassing only the OPTIONS method, pre-flight requests are successful and then everything is fine (for example, POST is correctly handled by InfluxDB, as it returns Access-Control-* headers as expected, without any nginx intervention).

**Comment by sranka** [comment]
@LaXiS96 thank you for letting us know. Please add 'User-Agent'  to Access-Control-Allow-Headers for Safari and Firefox to work properly.

**Comment by Fredo70** [comment]
I had the same problem. I am not very experienced with nginx. But it didn't work right away. I had to add "always" to Access-Control-Allow-Origin. Then it works.
`add_header 'Access-Control-Allow-Origin' '*' always;`

**Comment by sranka** [comment]
This is already fixed in v1.9.3.
